File: Current.py
# -*- coding: utf-8 -*-
"""
Created on Sat Dec  9 02:47:54 2023

@author: ASUS
"""


import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import LinearSegmentedColormap

# Solving the Laplace Equation for Obtaining Streamline Field in a 2D Forward Step
# Specifying parameters
x1 = 95.0
x2 = 57.0
h1 = 45
h2 = 27
time = 200


q = 45.0
u_inlet = q / (h1 * 1.0)
u_outlet = q / (h2 * 1.0)

nx = 153
ny = 46

del_y = h1 / (ny - 1)
del_x = (x1 + x2) / (nx - 1)
beta = (del_x / del_y) ** 2

# Grid
y = np.linspace(0, h1, ny).reshape(-1, 1)
y = np.tile(y, (1, nx))

x = np.linspace(0, x1 + x2, nx).reshape(1, -1)
x = np.tile(x, (ny, 1))

# Applying initial value to variables
error = 1.0

initial_s = np.zeros((ny, nx))
final_s = np.zeros((ny, nx))
ekhtelaf_s = np.zeros((ny, nx))
u = np.zeros((ny, nx))
v = np.zeros((ny, nx))

# Inlet Boundary Condition
for i in range(1, ny):
    initial_s[i, 0] = initial_s[i - 1, 0] + (u_inlet * del_y)
    final_s[i, 0] = final_s[i - 1, 0] + (u_inlet * del_y)

initial_s[-1, :] = initial_s[-1, 0]
final_s[-1, :] = final_s[-1, 0]

# Outlet Boundary Condition
for i in range(ny - 1, int(ny - (h2 / del_y)) - 1, -1):
    initial_s[i - 1, -1] = initial_s[i, -1] - (u_outlet * del_y)
    final_s[i - 1, -1] = final_s[i, -1] - (u_outlet * del_y)

# Iterative solving the streamline field
while error > 1e-10:
    initial_s = final_s.copy()

    for i in range(1, ny - 1):
        for j in range(1, nx - 1):
            if x[i, j] < x1:
                final_s[i, j] = (
                    initial_s[i, j + 1]
                    + initial_s[i, j - 1]
                    + beta * (initial_s[i - 1, j] + initial_s[i + 1, j])
                ) / (2 * (1 + beta))
            elif x[i, j] >= x1 and y[i, j] > (h1 - h2):
                final_s[i, j] = (
                    initial_s[i, j + 1]
                    + initial_s[i, j - 1]
                    + beta * (initial_s[i - 1, j] + initial_s[i + 1, j])
                ) / (2 * (1 + beta))

    ekhtelaf_s = final_s - initial_s
    error = np.max(np.abs(ekhtelaf_s))

# Calculating the velocity field
u = np.zeros((ny, nx))
v = np.zeros((ny, nx))
u[1:, 0] = u_inlet

for i in range(1, ny):
    for j in range(1, nx):
        u[i, j] = (final_s[i, j] - final_s[i - 1, j]) / del_y
        v[i, j] = -(final_s[i, j] - final_s[i, j - 1]) / del_x

# Exporting output for Tecplot
a = np.zeros(((nx) * (ny), 4))

k = 0
for i in range(ny):
    for j in range(nx):
        a[k, 0] = x[i, j]
        a[k, 1] = y[i, j]
        a[k, 2] = u[i, j]
        a[k, 3] = v[i, j]
        k += 1


# Handle non-finite values in x, y, and u
x = np.ma.masked_invalid(x).filled(np.nan)
y = np.ma.masked_invalid(y).filled(np.nan)
u = np.ma.masked_invalid(u).filled(np.nan)

# Create a pseudocolor plot
pcm = plt.pcolormesh(x, y, u, cmap='jet')

# Add labels and title
plt.xlabel('X-axis')
plt.ylabel('Y-axis')
plt.title('Velocity Field')

# Add colorbar for reference
plt.colorbar(pcm)

# Adjust the aspect ratio of the axes for a better representation
plt.axis('equal')
plt.axis('tight')

# Display the grid (optional)
plt.grid(False)

File: Current.m
clearvars 
clc
close all
% Solving The Laplace Equation For  Obtaining Streamline Field in a 2D Forward Step
% Specifying parameters
X1 = 95.0 ;
X2 = 57.0 ;
H1 = 45 ;
H2 = 27 ;
Q = 45.0 ;
U_INLET = Q/(H1*1.0) ;
U_OUTLET = Q/(H2*1.0) ;
 
NX = 153;
NY = 46 ;
 
 
DEL_Y = H1/(NY-1) ;
DEL_X = (X1+X2)/(NX-1) ;
BETA = (DEL_X/DEL_Y).^2 ;
clearvars 
clc
close all
% Solving The Laplace Equation For  Obtaining Streamline Field in a 2D Forward Step
% Specifying parameters
X1 = 95.0 ;
X2 = 57.0 ;
H1 = 45 ;
H2 = 27 ;
Q = 45.0 ;
U_INLET = Q/(H1*1.0) ;
U_OUTLET = Q/(H2*1.0) ;
 
NX = 153;
NY = 46 ;
 
 
DEL_Y = H1/(NY-1) ;
DEL_X = (X1+X2)/(NX-1) ;
BETA = (DEL_X/DEL_Y).^2 ;

% Applying Initial Value to Variables
ERROR = 1.0 ;
 
INITIAL_S = zeros(NY,NX) ;
FINAL_S = zeros(NY,NX) ;
EKHTELAF_S = zeros(NY,NX) ;
U = zeros(NY,NX) ;
V = zeros(NY,NX) ;

% Grid
Y = 0 : DEL_Y : H1 ;
Y = Y' ;
Y = repmat (Y ,1 , NX);
 
X = 0 : DEL_X : (X1+X2) ;
X = repmat(X,NY,1) ;

% Inlet Boundary Condition
for I = 2 : NY 
    INITIAL_S(I , 1) = INITIAL_S(I-1,1)+(U_INLET*DEL_Y) ;
    FINAL_S(I,1) = FINAL_S(I-1,1)+(U_INLET*DEL_Y) ;
end
% Top Boundary Condition
INITIAL_S(NY , :) = INITIAL_S(NY , 1) ;
FINAL_S(NY , :) = FINAL_S(NY , 1) ;
 
 
% Outlet Boundary Condition
for I = NY : -1 : ((NY -(H2/DEL_Y))+1)
    
    INITIAL_S(I-1 , NX) = INITIAL_S(I , NX)-(U_OUTLET*DEL_Y) ;
    FINAL_S(I-1 , NX) = FINAL_S(I , NX)-(U_OUTLET*DEL_Y) ;
end

% Iterative Solving the Streamline Field
while ERROR > 1e-10
    
    INITIAL_S = FINAL_S ;
    
    for I = 2 : NY-1 
        for J = 2 : NX-1
            
            if X(I,J) < X1 
                FINAL_S(I,J) = (INITIAL_S(I,J+1)+INITIAL_S(I,J-1)+...
                    (BETA*(INITIAL_S(I-1,J)+INITIAL_S(I+1,J))))/(2*(1+BETA)) ;
                
            else if (X(I,J) >= X1 && Y(I,J) > (H1-H2))
                    
                    FINAL_S(I,J) = (INITIAL_S(I,J+1)+INITIAL_S(I,J-1)+...
                        (BETA*(INITIAL_S(I-1,J)+INITIAL_S(I+1,J))))/(2*(1+BETA)) ;
                end
            end
        end
end
 
 
EKHTELAF_S = FINAL_S - INITIAL_S ;
 
ERROR = max(max(abs(EKHTELAF_S))) ;
 
end
% Calculating the Velocity Field
 U(2:NY,1) = U_INLET ;
 
for I = 2 : NY
   for J = 2 : NX
       
    U(I,J) = (FINAL_S(I,J)-FINAL_S(I-1,J))/DEL_Y ;
    V(I,J) = -(FINAL_S(I,J)-FINAL_S(I,J-1))/DEL_X ; 
    
   end
end
% Exporting Output for Tecplot
for i = 1 : (NX)*(NY)
    
            a(i,1)=X(i);
            a(i,2)=Y(i);
            a(i,3)=U(i);
            a(i,4)=V(i);
            
end
% Find indices where U is zero
zero_indices = U == 0;

% Set the corresponding values in X, Y, and U to NaN
X(zero_indices) = NaN;
Y(zero_indices) = NaN;
U(zero_indices) = NaN;

% Assuming X, Y, and U are already defined as 46x153 matrices

figure(1)
% Create a 3D surface plot
surf(X, Y, U, 'EdgeColor', 'none');

% Set colormap to jet
colormap(turbo(256));

% Add labels and title
xlabel('X (m)');
ylabel('Y (m)');
zlabel('Velocity (U)');
title('Vertical Velocity');

xticks(0:4:152)

% Add colorbar for reference
clim([0 3])
colorbar;
axis equal tight;

% Adjust the view for better visualization (optional)
view(2); % 2D view from the top

grid off;

figure(2)
[startX,startY] = meshgrid(0,0:1:45);
h = streamline(0:1:152, 0:1:45, U, V, startX, startY);
set(h, 'Color', 'black', 'LineWidth', 1.5); % Customize streamline appearance
axis equal tight;
xticks(0:4:152)
title('Streamlines')
xlabel('X (m)')
ylabel('Y (m)')




